lxd/container_lxc.go

```diff
@@ -8046,21 +8099,40 @@ func (c *containerLXC) insertNetworkDevice(name string, m types.Device) (types.D
 		return nil, err
 	}
 
-	// Create the interface
-	devName, err := c.createNetworkDevice(name, m)
-	if err != nil {
-		return nil, err
-	}
+	// Setup network device if not type ipvlan (which is done via liblxc only)
+	if m["nictype"] != "ipvlan" {
+		err := c.checkIPVLANSupport()
+		if err != nil {
+			return nil, err
+		}
 
-	// Add the interface to the container
-	err = c.c.AttachInterface(devName, m["name"])
-	if err != nil {
-		return nil, fmt.Errorf("Failed to attach interface: %s: %s", devName, err)
+		// Create the interface
+		devName, err := c.createNetworkDevice(name, m)
+		if err != nil {
+			return nil, err
+		}
+
+		// Add the interface to the container
+		err = c.c.AttachInterface(devName, m["name"])
+		if err != nil {
+			return nil, fmt.Errorf("Failed to attach interface: %s: %s", devName, err)
+		}
 	}
 
 	return m, nil
 }
 
+// checkIPVLANSupport checks whether the liblxc available has the necessary IPVLAN features.
+func (c *containerLXC) checkIPVLANSupport() error {
+	if lxc.HasApiExtension("network_ipvlan") &&
+		lxc.HasApiExtension("network_l2proxy") &&
+		lxc.HasApiExtension("network_gateway_device_route") {
```

Comment:
With `lxc_features` moving this to the State struct, we can perform that check without actually hitting liblxc.